### PR TITLE
Free lockable-resource assigned to dead builds

### LINKED ISSUE #500: release locks on restart
### What feature do you want to see added?

Hi

Currently some of the pipelines mentions a previous build is using the lock, where the previous build is already complete.
This happens usually after Jenkins crashes or when forced restart.

### Upstream changes

_No response_

### PR DESCRIPTION
**Pull request by mPokornyETM**
#fix #500 

Sometimes after re-starts (jenkins crashed or what ever) are resources still locked by build, but the build is no more running.
This changes will 'unlock' all resource assigned to the dead builds

### Testing done

It is more or less not possible to simulate it.
We use this code in our Jenkins instance few weeks and it works
We reboot our instances once per week (or often) and after this changes we has the issue no more.
Therefore I thing it works as expected.

### Proposed upgrade guidelines

N/A

### Localizations

N/A

### Submitter checklist

- [x] The Jira / Github issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - The changelog generator for plugins uses the **pull request title as the changelog entry**.
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during the upgrade.
- [x] There is automated testing or an explanation that explains why this change has no tests.
- ~~[ ] New public functions for internal use only are annotated with `@NoExternalUse`. In case it is used by non java code the `Used by {@code <panel>.jelly}` Javadocs are annotated.~~
- ~~[ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease the future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).~~
- ~~[ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.~~
- ~~[ ] For new APIs and extension points, there is a link to at least one consumer.~~
- ~~[ ] Any localizations are transferred to *.properties files.~~
- ~~[ ] Changes in the interface are documented also as [examples](src/doc/examples/readme.md).~~

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There is at least one (1) approval for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the **pull request title** and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically. See also [release-drafter-labels](https://github.com/jenkinsci/.github/blob/ce466227c534c42820a597cb8e9cac2f2334920a/.github/release-drafter.yml#L9-L50).
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] java code changes are tested by automated test.

### CODE REVIEW

**Comment by jimklimov** [comment]
Just in case: would this fire on "server start", or on "job load" (per annotation name) so e.g. when an MBP job gets defined due to new branch, or otherwise a job configuration edited (and so reloaded?)

I have concern for custom uses of Lockable Resources (sort of hackery beside defined steps) which toss a particular resource like a SUT between stages or even jobs.

File: src/main/java/org/jenkins/plugins/lockableresources/FreeDeadJobs.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import java.util.logging.Logger;
+import org.jenkins.plugins.lockableresources.queue.LockableResourcesStruct;
+import org.jenkinsci.plugins.workflow.steps.StepContext;
+
+/**
+ * Sometimes after re-starts (jenkins crashed or what ever) are resources still
+ * locked by build, but the build is no more running.
+ * This script will 'unlock' all resource assigned to dead builds
+ */
+@Deprecated
```
Comment (mPokornyETM) [inline src/main/java/org/jenkins/plugins/lockableresources/FreeDeadJobs.java:27]: this is wrong

**Comment by mPokornyETM** [comment]
> Just in case: would this fire on "server start", or on "job load" (per annotation name) so e.g. when an MBP job gets defined due to new branch, or otherwise a job configuration edited (and so reloaded?)
> 
> I have concern for custom uses of Lockable Resources (sort of hackery beside defined steps) which toss a particular resource like a SUT between stages or even jobs.

Currently on on jenkins start - excactly 'InitMilestone.JOB_LOADED' see also
https://javadoc.jenkins.io/hudson/init/InitMilestone.html#JOB_LOADED

I make similar loic in your Jenkins instance before we try to lock someone resource. But it might be dangeraus here.

File: src/main/java/org/jenkins/plugins/lockableresources/FreeDeadJobs.java (hunk 1 above)
Comment (g3n35i5) [reply src/main/java/org/jenkins/plugins/lockableresources/FreeDeadJobs.java:27]: It isn't deprecated, right? :laughing:

Comment (g3n35i5) [inline src/main/java/org/jenkins/plugins/lockableresources/FreeDeadJobs.java:27]:
```suggestion
 */
```

**Review by g3n35i5 — CHANGES_REQUESTED** [review]

**Review by g3n35i5 — APPROVED** [review]
LGTM. I think what you *could* do for testing this is hard-killing jenkins (SIGKILL, kill -9) during a job run?

**Comment by nrayapati** [comment]
This change is breaking the Jenkins restarts. 

@g3n35i5 @mPokornyETM 

Screenshot 2023-08-25 at 3 11 20 AM

**Comment by mPokornyETM** [comment]
see #550

**Pull request merged**